RedundantModifierCheck throws NullPointerException

Check documentation: 
[RedundantModifier](https://checkstyle.sourceforge.io/config_modifier.html#RedundantModifier)


````
$ cat TestClass.java
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.Callable;

public class DaytimeTask implements Runnable, Callable<Void> {
  private static final String UTF_8 = StandardCharsets.UTF_8.name();
  private final Socket connection;

  DaytimeTask(Socket connection) {
    this.connection = connection;
  }

  @Override
  public Void call() {
    run();
    return null;
  }

  @Override
  @SuppressWarnings("PMD.LawOfDemeter")
  public void run() {
    try (Writer out =
             new OutputStreamWriter(connection.getOutputStream(), UTF_8);
         connection;) {
      Date now = new Date();
      SimpleDateFormat format =
          new SimpleDateFormat("yy-MM-dd hh:mm:ss Z", Locale.getDefault());
      out.write(ProcessHandle.current().pid() + " " + format.format(now)
                + "\r\n");
      out.flush();
    } catch (IOException ex) {
      System.err.println(ex);
    }
  }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
    <module name="RedundantModifier" />
    </module>
</module>

$ java -jar checkstyle-8.33-all.jar -c TestConfig.xml TestClass.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing TestClass.java
    at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:311)
    at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
    at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
    at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
    at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
    at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: java.lang.NullPointerException
    at com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck.checkForRedundantModifier(RedundantModifierCheck.java:424)
    at com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck.processResources(RedundantModifierCheck.java:413)
    at com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck.visitToken(RedundantModifierCheck.java:237)
    at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:340)
    at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:451)
    at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:278)
    at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:151)
    at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
    at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
    at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:298)
    ... 5 more
Checkstyle ends with 1 errors.
````


---------------

Describe what you expect in detail.

Checkstyle should complete successfully either emitting violations or no error messages when there are none. No NullPointerException as above must occur.

--------------

First reported at Checkstyle Google group:

https://groups.google.com/forum/m/#!topic/checkstyle/OZvmwJypxvI

Links to additional files throwing similar exception in Google Group post.

This is the output in my errors output file for the same with the full configuration that includes all suppression filters.

```
Starting audit...
[ERROR] /data/data/com.termux/files/home/LearnJava/TestCode/./cstests/DaytimeTask.java:1: Got an exception - java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck.checkForRedundantModifier(RedundantModifierCheck.java:424)
	at com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck.processResources(RedundantModifierCheck.java:413)
	at com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck.visitToken(RedundantModifierCheck.java:237)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:340)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:451)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:278)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:151)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:298)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
 [Checker]
Audit done.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
@@ -21,12 +21,14 @@ package com.puppycrawl.tools.checkstyle.checks.modifier;
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
 
 import com.puppycrawl.tools.checkstyle.StatelessCheck;
 import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
  * <p>
@@ -420,15 +422,12 @@ public class RedundantModifierCheck
      * @param modifierType The modifier to check for.
      */
     private void checkForRedundantModifier(DetailAST ast, int modifierType) {
-        final DetailAST astModifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
-        DetailAST astModifier = astModifiers.getFirstChild();
-        while (astModifier != null) {
-            if (astModifier.getType() == modifierType) {
-                log(astModifier, MSG_KEY, astModifier.getText());
-            }
-
-            astModifier = astModifier.getNextSibling();
-        }
+        Optional.ofNullable(ast.findFirstToken(TokenTypes.MODIFIERS))
+            .ifPresent(modifiers -> {
+                TokenUtil.forEachChild(modifiers, modifierType, modifier -> {
+                    log(modifier, MSG_KEY, modifier.getText());
+                });
+            });
     }
 
     /**

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
@@ -270,27 +270,9 @@ public class RedundantModifierCheck
      */
     private void checkEnumConstructorModifiers(DetailAST ast) {
         final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
-        final DetailAST modifier = getFirstModifierAst(modifiers);
-
-        if (modifier != null) {
-            log(modifier, MSG_KEY, modifier.getText());
-        }
-    }
-
-    /**
-     * Retrieves the first modifier that is not an annotation.
-     *
-     * @param modifiers The ast to examine.
-     * @return The first modifier or {@code null} if none found.
-     */
-    private static DetailAST getFirstModifierAst(DetailAST modifiers) {
-        DetailAST modifier = modifiers.getFirstChild();
-
-        while (modifier != null && modifier.getType() == TokenTypes.ANNOTATION) {
-            modifier = modifier.getNextSibling();
-        }
-
-        return modifier;
+        TokenUtil.findFirstTokenByPredicate(
+            modifiers, mod -> mod.getType() != TokenTypes.ANNOTATION
+        ).ifPresent(modifier -> log(modifier, MSG_KEY, modifier.getText()));
     }
 
     /**
@@ -385,13 +367,9 @@ public class RedundantModifierCheck
      */
     private void processAbstractMethodParameters(DetailAST ast) {
         final DetailAST parameters = ast.findFirstToken(TokenTypes.PARAMETERS);
-
-        for (DetailAST child = parameters.getFirstChild(); child != null; child = child
-                .getNextSibling()) {
-            if (child.getType() == TokenTypes.PARAMETER_DEF) {
-                checkForRedundantModifier(child, TokenTypes.FINAL);
-            }
-        }
+        TokenUtil.forEachChild(parameters, TokenTypes.PARAMETER_DEF, paramDef -> {
+            checkForRedundantModifier(paramDef, TokenTypes.FINAL);
+        });
     }
 
     /**
@@ -526,13 +504,7 @@ public class RedundantModifierCheck
     private static List<DetailAST> getMethodAnnotationsList(DetailAST methodDef) {
         final List<DetailAST> annotationsList = new ArrayList<>();
         final DetailAST modifiers = methodDef.findFirstToken(TokenTypes.MODIFIERS);
-        DetailAST modifier = modifiers.getFirstChild();
-        while (modifier != null) {
-            if (modifier.getType() == TokenTypes.ANNOTATION) {
-                annotationsList.add(modifier);
-            }
-            modifier = modifier.getNextSibling();
-        }
+        TokenUtil.forEachChild(modifiers, TokenTypes.ANNOTATION, annotationsList::add);
         return annotationsList;
     }
 

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
@@ -262,4 +262,14 @@ public class RedundantModifierCheckTest
                 expected);
     }
 
+    @Test
+    public void testTryWithResourcesBlock() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(RedundantModifierCheck.class);
+        final String[] expected = {
+            "10:19: " + getCheckMessage(MSG_KEY, "final"),
+        };
+        verify(checkConfig, getNonCompilablePath("InputRedundantModifierTryWithResources.java"),
+                expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierTryWithResources.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierTryWithResources.java
@@ -1,0 +1,16 @@
+//non-compiled with javac: Compilable with Java9
+package com.puppycrawl.tools.checkstyle.checks.modifier.redundantmodifier;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+
+public class InputRedundantModifierTryWithResources {
+    public static void main(String[] args) {
+        OutputStreamWriter out = null;
+        try (out; final OutputStreamWriter out2 = null;) {
+            out.write(1);
+        } catch (IOException ex) {
+            System.err.println(ex);
+        }
+    }
+}

